File: rl/engine/deck.py
import random
from typing import List

from . import constants
from .card import Card, Suit


class Deck:
    def __init__(self) -> None:
        self.cards: List[Card] = [
            Card(suit, rank) for suit in Suit for rank in range(1, constants.MAX_CARD_RANK + 1)
        ]

    def shuffle(self) -> None:
        random.shuffle(self.cards)

    def deal(self, num_players: int, cards_per_player: int) -> List[List[Card]]:
        return [
            self.cards[i * cards_per_player: (i + 1) * cards_per_player]
            for i in range(num_players)
        ]

    def reset(self) -> None:
        self.__init__()

File: engine/src/main/java/com/mjaseem/trick/engine/Deck.java
package com.mjaseem.trick.engine;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// Deck of cards
public class Deck {
    public static final int MAX_CARD_RANK = 14;
    private final List<Card> cards = new ArrayList<>();

    public Deck() {
        for (Records.Suit suit : Records.Suit.values()) {
            for (int i = 1; i <= MAX_CARD_RANK; i++) {
                cards.add(new Card(suit, i));
            }
        }
    }

    public void shuffle() {
        Collections.shuffle(cards);
    }

    public List<List<Card>> deal(int players, int cardsPerPlayer) {
        List<List<Card>> hands = new ArrayList<>();
        for (int i = 0; i < players; i++) {
            hands.add(new ArrayList<>(cards.subList(i * cardsPerPlayer, (i + 1) * cardsPerPlayer)));
        }
        return hands;
    }

    public void reset() {
        cards.clear();
        for (Records.Suit suit : Records.Suit.values()) {
            for (int i = 1; i <= 14; i++) {
                cards.add(new Card(suit, i));
            }
        }
    }
}
